bench/src/main/scala/cats/bench/TraverseBench.scala

```diff
@@ -21,155 +21,182 @@
 
 package cats.bench
 
-import cats.{Always, Applicative, Eval, Traverse}
-import cats.instances.either._
-import cats.instances.list._
-import cats.instances.vector._
-import org.openjdk.jmh.annotations.{Benchmark, Scope, State}
+import cats.{Eval, Traverse, TraverseFilter}
+import org.openjdk.jmh.annotations.{Benchmark, Param, Scope, Setup, State}
+import org.openjdk.jmh.infra.Blackhole
 
 @State(Scope.Benchmark)
-class TraverseListBench {
-  val instance: Traverse[List] = Traverse[List]
-  val f: Int => Either[Int, Int] = Right(_)
-
-  val xs1: List[Int] = (1 to 10).toList
-  val xs2: List[Int] = (1 to 100).toList
-  val xs3: List[Int] = (1 to 1000).toList
-  val xs4: List[Int] = (1 to 10000).toList
-
-  def traverseCats[G[_], A, B](fa: List[A])(f: A => G[B])(implicit G: Applicative[G]): G[List[B]] =
-    instance
-      .foldRight[A, G[List[B]]](fa, Always(G.pure(List.empty))) { (a, lglb) =>
-        G.map2Eval(f(a), lglb)(_ :: _)
+class TraverseBench {
+  val listT: Traverse[List] = Traverse[List]
+  val listTFilter: TraverseFilter[List] = TraverseFilter[List]
+
+  val vectorT: Traverse[Vector] = Traverse[Vector]
+  val vectorTFilter: TraverseFilter[Vector] = TraverseFilter[Vector]
+
+  // the unit of CPU work per iteration
+  private[this] val Work: Long = 10
+
+  private[this] case object Failure extends RuntimeException
+
+  @Param(Array("10000"))
+  var length: Int = _
+
+  var list: List[Int] = _
+  var vector: Vector[Int] = _
+
+  @Setup
+  def setup(): Unit = {
+    list = 0.until(length).toList
+    vector = 0.until(length).toVector
+  }
+
+  @Benchmark
+  def traverseList(bh: Blackhole) = {
+    val result = listT.traverse(list) { i =>
+      Eval.later {
+        Blackhole.consumeCPU(Work)
+        i * 2
       }
-      .value
+    }
+
+    bh.consume(result.value)
+  }
 
-  def traverseFoldRight[G[_], A, B](fa: List[A])(f: A => G[B])(implicit G: Applicative[G]): G[List[B]] =
-    fa.foldRight[Eval[G[List[B]]]](Always(G.pure(Nil))) { case (h, t) =>
-      G.map2Eval(f(h), Eval.defer(t))(_ :: _)
-    }.value
+  @Benchmark
+  def traverseListError(bh: Blackhole) = {
+    val result = listT.traverse(list) { i =>
+      Eval.later {
+        Blackhole.consumeCPU(Work)
 
-  def traverseRec[G[_], A, B](fa: List[A])(f: A => G[B])(implicit G: Applicative[G]): G[List[B]] = {
-    def loop(fa: List[A]): Eval[G[List[B]]] = fa match {
-      case h :: t => G.map2Eval(f(h), Eval.defer(loop(t)))(_ :: _)
-      case Nil    => Eval.now(G.pure(Nil))
+        if (i == length * 0.3) {
+          throw Failure
```

Comment:
I'm surprised to see us using `throw` in these benchmarks along with `Eval`. Can you add a comment since it is non-idiomatic to do so?

Reply:
`Eval` secretly supports short-circuiting, which means it's a halfway-decent approximate for `IO`. Ultimately, I would be willing to bet that 90% of `traverse` usage in the wild has `IO` as the base applicative. In this case, I was just trying to caliper down on the short-circuiting scenario without adding extra overhead that would muddy the constant factors.

The other advantage to doing it this way is it allows us to compare like-for-like. If we write the same `traverse` process by hand (which I have now done locally in Cats Effect), we can compare directly and see the overhead of `traverse` itself, allowing us to explore different strategies.